protocols/identify/src/behaviour.rs

```diff
@@ -519,4 +549,38 @@ mod tests {
         ));
         assert!(multiaddr_matches_peer_id(&addr_without_peer_id, &peer_id));
     }
+
+    #[test]
+    fn filter_new_external_addrs_leaves_new_addrs() {
+        let addr1 = Multiaddr::empty().with(multiaddr::Protocol::Memory(rand::random::<u64>()));
+        let addr2 = Multiaddr::empty().with(multiaddr::Protocol::Memory(rand::random::<u64>()));
+
+        let old_addrs: Vec<Multiaddr> = [0..3]
+            .iter()
+            .map(|_x| Multiaddr::empty().with(multiaddr::Protocol::Memory(rand::random::<u64>())))
+            .collect();
+
+        let new_addrs = Vec::from([addr1, addr2]);
+
+        assert_eq!(
+            filter_new_external_addrs(&old_addrs, &new_addrs),
+            Some(new_addrs.clone()),
+        );
+    }
+
+    #[test]
+    fn filter_new_external_addrs_filter_duplicates() {
+        let addr1 = Multiaddr::empty().with(multiaddr::Protocol::Memory(rand::random::<u64>()));
+        let addr2 = Multiaddr::empty().with(multiaddr::Protocol::Memory(rand::random::<u64>()));
+
+        let mut old_addrs: Vec<Multiaddr> = [0..3]
+            .iter()
+            .map(|_x| Multiaddr::empty().with(multiaddr::Protocol::Memory(rand::random::<u64>())))
+            .collect();
+
+        let new_addrs = Vec::from([addr1.clone(), addr2.clone()]);
+        old_addrs.append(&mut new_addrs.clone());
+
+        assert_eq!(filter_new_external_addrs(&old_addrs, &new_addrs), None)
+    }
```

Comment:
I think I'd rather not like to rely on randomness here. Hardcoded numbers should make this easier to reason about. Plus, you can then make a utility function for giving you a `/memory` address where you just pass a `u64` in. That should shorten these tests a lot.